anndata/experimental/read_remote/read_remote.py

```diff
@@ -0,0 +1,335 @@
+from collections import OrderedDict, abc as cabc
+from copy import copy
+from functools import cached_property
+from pathlib import Path
+from typing import Any, MutableMapping, Union, List, Sequence
+from anndata._core.aligned_mapping import Layers, PairwiseArrays
+from anndata._core.anndata import StorageType, _check_2d_shape, _gen_dataframe
+from anndata._core.file_backing import AnnDataFileManager
+from anndata._core.index import Index
+from anndata._core.raw import Raw
+from anndata._core.views import _resolve_idxs
+from anndata._io.specs.registry import read_elem
+from anndata.compat import _move_adj_mtx, _read_attr
+from anndata.utils import convert_to_dict
+
+import zarr
+import pandas as pd
+import numpy as np
+
+from ..._core import AnnData, AxisArrays
+from .. import read_dispatched
+
+
+# TODO: Do we really need to subclass the Array class here?  Ryan Abernathy seems to say "no"
+# but I don't really want to mess with the methods.  The downside is that (for some reason), it's
+# reading the `zarray` of the `codes` path which should not have to happen, but I can't figure out a way around it.
+class CategoricalZarrArray(zarr.core.Array):
+    def __init__(self, group, *args, **kwargs):
+        codes_path = group.path + "/codes"
+        super().__init__(group.store.store, codes_path, *args, **kwargs)
+        self._categories = group["categories"]
+        self._group_attrs = group.attrs
+
+    @cached_property
+    def categories(self):
+        return self._categories[()]
+
+    @cached_property
+    def ordered(self):
+        return bool(_read_attr(self._group_attrs, "ordered"))
+
+    def __array__(self, *args):  # may need to override this, copied for now
+        a = self[...]
+        if args:
+            a = a.astype(args[0])
+        return a
+
+    def __getitem__(self, selection):
+        result = super().__getitem__(selection)
+        return pd.Categorical.from_codes(
+            codes=result,
+            categories=self.categories,
+            ordered=self.ordered,
+        )
```

Comment:
Looking again, I think you more or less want to do what was suggested. It's just a very minimal implementation of an array, that supports indexing.

I think the zarr array class is not appropriate to subclass like this. It has a ton of assumptions about how it's structured, which just aren't met by the categorical encoding here.
```suggestion
import numpy as np
import pandas as pd

# Doesn't need to use this, just using it for a couple inherited methods
from xarray.core.indexing import ExplicitlyIndexedNDArrayMixin

class LazyCategoricalArray(ExplicitlyIndexedNDArrayMixin):
    __slots__ = ("codes", "categories", "attrs")

    def __init__(self, group, *args, **kwargs):
        self.codes = group["codes"]
        self.categories = group["categories"][...]  # slots don't mix with cached_property, ExpicitlyIndexedArray uses slots
        self.attrs = dict(group.attrs)

    @property
    def dtype(self) -> pd.CategoricalDtype:
        return pd.CategoricalDtype(self.categories, self.ordered)
    
    @property
    def shape(self) -> tuple[int, ...]:
        return self.codes.shape

    @property
    def ordered(self):
        return bool(self.attrs["ordered"])

    def __array__(self, *args) -> np.ndarray:  # may need to override this, copied for now
        a = self[...]
        if args:
            a = a.astype(args[0])
        return a

    def __getitem__(self, selection) -> pd.Categorical:
        return pd.Categorical.from_codes(
            codes=self.codes.oindex[selection],
            categories=self.categories,
            ordered=self.ordered,
        )

    def __repr__(self) -> str:
        return f"LazyCategoricalArray(codes=..., categories={self.categories}, ordered={self.ordered})"
```

Reply:
Sounds good.  Wasn't a huge fan of adding another dependency, but I guess `xarray` is coming at some point anyway.

Reply:
Could also do it without xarray, though it's a very light dependency.

Reply:
I think you'd just need to define things like `__length__`, `.ndim`, etc.